### PR TITLE
Introduce `show` for `TimeEvolutionSol` and `TimeEvolutionMCSol`, also, adjust return `states` and `expect` from `sesolve`, `mesolve`, and `mcsolve`

### LINKED ISSUE #143: `expect` type after solving dynamics
# `sol.states`
In `qutip`, the results is stored depends on `e_ops`:
- if `e_ops` is specified: don't return states at each time point (even the final state, so `sol.states = []`)
- if `e_ops` is not specified: return states at each time point

I think we should also do the same thing.

However, if the keyword argument `saveat` is given, we can still store the state even `e_ops` is given

### PR DESCRIPTION
**Pull request by ytdHuang**
close #143 

Summary of this PR:
1. Adjust many docstrings for time evolution functions (add `@doc raw`, LaTeX formula, and some minor changes)
2. Introduce `show` for `TimeEvolutionSol` and `TimeEvolutionMCSol`: take the brief example in README, here is the output if we `show` (or `print`) `sol` directly:
```
julia> sol = mesolve(H, ψ0, tlist, c_ops, e_ops = e_ops)
Solution of time evolution
(return code: Success)
--------------------------
num_states = 1
num_expect = 1
ODE alg.: Tsit5(; stage_limiter! = trivial_limiter!, step_limiter! = trivial_limiter!, thread = static(false),)
abstol = 1.0e-7
reltol = 1.0e-5
```
3. Adjust the return data for `sesolve`, `mesolve`, and `mcsolve`

For the current implementation, the users will only get a final state if they doesn't specify `e_ops`, which cause some ambiguity about the argument `t_l`. In this PR, the `expect` will be decide by the argument `e_ops`, and the `states` will be decide by the argument `saveat`:
- if the user specifies only `e_ops`: `saveat = [t_l[end]]`, thus, return expectation values and the final state (which is same as current implementation)
- if the user doesn't specifies `e_ops`: `saveat = t_l`, thus, return all the states corresponding to `t_l`
- if the user specifies `e_ops` and `saveat` at the same time: `expect` will depend on `e_ops`, and `states` will depend on `saveat`.

Again, take the brief example in README:

if user doesn't specify `e_ops`:
```
julia> sol = mesolve(H, ψ0, tlist, c_ops)
Solution of time evolution
(return code: Success)
--------------------------
num_states = 100
num_expect = 0
ODE alg.: Tsit5(; stage_limiter! = trivial_limiter!, step_limiter! = trivial_limiter!, thread = static(false),)
abstol = 1.0e-7
reltol = 1.0e-5
```

if user specifies both `e_ops` and `saveat`:
```
julia> sol = mesolve(H, ψ0, tlist, c_ops, e_ops = e_ops, saveat = tlist)
Solution of time evolution
(return code: Success)
--------------------------
num_states = 100
num_expect = 1
ODE alg.: Tsit5(; stage_limiter! = trivial_limiter!, step_limiter! = trivial_limiter!, thread = static(false),)
abstol = 1.0e-7
reltol = 1.0e-5
```
```
julia> sol = mesolve(H, ψ0, tlist, c_ops, e_ops = e_ops, saveat = [])
Solution of time evolution
(return code: Success)
--------------------------
num_states = 0
num_expect = 1
ODE alg.: Tsit5(; stage_limiter! = trivial_limiter!, step_limiter! = trivial_limiter!, thread = static(false),)
abstol = 1.0e-7
reltol = 1.0e-5
```

### CODE REVIEW

**Comment by albertomercurio** [comment]
Ok, yes. It makes sense that, when `e_ops` is empty, then it returns the states, yes. By the way, in the last example you made, I think that we should avoid to use `save at=[]`. As far as I remember, the option `saveat = []` will return all the states at each time step (different from `t_l`). See, e.g., [the documentation](https://docs.sciml.ai/DiffEqDocs/stable/basics/common_solver_opts/).

**Comment by ytdHuang** [comment]
> Ok, yes. It makes sense that, when `e_ops` is empty, then it returns the states, yes. By the way, in the last example you made, I think that we should avoid to use `save at=[]`. As far as I remember, the option `saveat = []` will return all the states at each time step (different from `t_l`). See, e.g., [the documentation](https://docs.sciml.ai/DiffEqDocs/stable/basics/common_solver_opts/).

The default of `saveat` will only be either `[t_l[end]]` or `t_l`.
It's just that we provide a degree of freedom for users to specify the `saveat`. But I think seldom people will specify this keyword argument.

**Comment by ytdHuang** [comment]
some outputs of `show` function for `TimeEvolutionMCSol`:

(again, take Brief example in README as demonstration):

(without specifying `e_ops`)
```julia
sol = mcsolve(H, ψ0, tlist, c_ops, n_traj = 2)
```
```
Solution of quantum trajectories
(converged: true)
--------------------------------
num_trajectories = 2
num_states = 100
num_expect = 0
ODE alg.: Tsit5(; stage_limiter! = trivial_limiter!, step_limiter! = trivial_limiter!, thread = static(false),)
abstol = 1.0e-7
reltol = 1.0e-5
```

(specify `e_ops`)
```julia
sol = mcsolve(H, ψ0, tlist, c_ops, n_traj = 2, e_ops = e_ops)
```
```
Solution of quantum trajectories
(converged: true)
--------------------------------
num_trajectories = 2
num_states = 1
num_expect = 1
ODE alg.: Tsit5(; stage_limiter! = trivial_limiter!, step_limiter! = trivial_limiter!, thread = static(false),)
abstol = 1.0e-7
reltol = 1.0e-5
```

(specify both `saveat` and `e_ops`)
```julia
sol = mcsolve(H, ψ0, tlist, c_ops, n_traj = 2, e_ops = e_ops, saveat = tlist)
```
```
Solution of quantum trajectories
(converged: true)
--------------------------------
num_trajectories = 2
num_states = 100
num_expect = 1
ODE alg.: Tsit5(; stage_limiter! = trivial_limiter!, step_limiter! = trivial_limiter!, thread = static(false),)
abstol = 1.0e-7
reltol = 1.0e-5
```

File: src/time_evolution/mcsolve.jl [hunk 1]
```diff
@@ -242,29 +268,52 @@ end
         output_func::Function=_mcsolve_output_func,
         kwargs...)
 
-Generates the ODEProblem for an ensemble of trajectories of the Monte Carlo wave function
-time evolution of an open quantum system.
+Generates the ODEProblem for an ensemble of trajectories of the Monte Carlo wave function time evolution of an open quantum system.
```
Comment (albertomercurio) [inline src/time_evolution/mcsolve.jl:271]: I would write “Generates the `EnsembleProblem` of `ODEProblem`s for the ensemble of trajectories of the Monte Carlo time evolution …”

File: src/time_evolution/mcsolve.jl [hunk 2]
```diff
@@ -242,29 +268,52 @@ end
...
+\frac{\partial}{\partial t} |\psi(t)\rangle= -i \hat{H}_{\textrm{eff}} |\psi(t)\rangle,
+```
+
+where
+
+```math
+\hat{H}_{\textrm{eff}} = \hat{H} - \frac{i}{2} \sum_i \hat{J}_i^\dagger \hat{J}_i.
+```
+
+During the time evolution, each quantum jump ``\hat{J}_i`` event happends at randomly determined time ``t^*`` can be described by
```
Comment (albertomercurio) [inline src/time_evolution/mcsolve.jl:285]: Same here, I would copy the QuTiP definition.

File: src/time_evolution/mcsolve.jl [hunk 3]
```diff
@@ -316,31 +365,53 @@ end
...
+```
 
-# Returns
+where
+
+```math
+\hat{H}_{\textrm{eff}} = \hat{H} - \frac{i}{2} \sum_i \hat{J}_i^\dagger \hat{J}_i.
+```
+
+During the time evolution, each quantum jump ``\hat{J}_i`` event happends at randomly determined time ``t^*`` can be described by
```
Comment (albertomercurio) [inline src/time_evolution/mcsolve.jl:380]: Same here.

Comment (ytdHuang) [reply src/time_evolution/mcsolve.jl]: Done~

File: src/time_evolution/mcsolve.jl (hunk 2 above)
Comment (ytdHuang) [reply src/time_evolution/mcsolve.jl]: Done~

File: src/time_evolution/mcsolve.jl (hunk 1 above)
Comment (ytdHuang) [reply src/time_evolution/mcsolve.jl]: Done~

File: src/time_evolution/mcsolve.jl [hunk 4]
```diff
@@ -145,12 +167,16 @@ function mcsolveProblem(
     jump_callback::TJC = ContinuousLindbladJumpCallback(),
     kwargs...,
 ) where {MT1<:AbstractMatrix,T2,Tc<:AbstractMatrix,Te<:AbstractMatrix,TJC<:LindbladJumpCallbackType}
-    H_eff = H - T2(0.5im) * mapreduce(op -> op' * op, +, c_ops)
+    H_eff = H - 0.5im * mapreduce(op -> op' * op, +, c_ops)
```
[the remark this answers is not in the document]
Comment (ytdHuang) [reply src/time_evolution/mcsolve.jl]: I have added it back

File: src/time_evolution/mcsolve.jl [hunk 5]
```diff
@@ -111,26 +109,50 @@ end
...
+\frac{\partial}{\partial t} |\psi(t)\rangle= -i \hat{H}_{\textrm{eff}} |\psi(t)\rangle,
+```
+
+where
+
+```math
+\hat{H}_{\textrm{eff}} = \hat{H} - \frac{i}{2} \sum_i \hat{J}_i^\dagger \hat{J}_i.
+```
+
+During the time evolution, each quantum jump ``\hat{J}_i`` event happends at randomly determined time ``t^*`` can be described by
```
[the remark this answers is not in the document]
Comment (ytdHuang) [reply src/time_evolution/mcsolve.jl]: Done~

**Comment by ytdHuang** [comment]
I've added the handling for unsupported keyword argument `save_idxs`, and also the corresponding runtests.

**Review by albertomercurio — APPROVED** [review]

**Pull request merged**